engine/crates/engine-v2/src/response/write/deserialize/selection_set/runtime_concrete.rs

```diff
@@ -0,0 +1,142 @@
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    fmt,
+};
+
+use serde::de::{DeserializeSeed, IgnoredAny, MapAccess, Visitor};
+
+use crate::{
+    plan::RuntimeCollectedSelectionSet,
+    response::{
+        write::deserialize::{key::Key, FieldSeed, SeedContextInner},
+        ResponseBoundaryItem, ResponseEdge, ResponseObject, ResponsePath, ResponseValue,
+    },
+};
+
+pub(crate) struct RuntimeConcreteCollectionSetSeed<'ctx, 'parent> {
+    pub ctx: &'parent SeedContextInner<'ctx>,
+    pub path: &'parent ResponsePath,
+    pub selection_set: &'parent RuntimeCollectedSelectionSet,
+}
+
+impl<'de, 'ctx, 'parent> DeserializeSeed<'de> for RuntimeConcreteCollectionSetSeed<'ctx, 'parent> {
+    type Value = ResponseValue;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_map(self)
+    }
+}
+
+impl<'de, 'ctx, 'parent> Visitor<'de> for RuntimeConcreteCollectionSetSeed<'ctx, 'parent> {
+    type Value = ResponseValue;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter.write_str("an object")
+    }
+
+    // later we could also support visit_struct by using the schema as the reference structure.
+    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+    where
+        A: MapAccess<'de>,
+    {
+        let plan = self.ctx.plan;
+        let keys = plan.response_keys();
+        let selection_set = self.selection_set;
+        let mut identifier = super::ObjectIdentifier::new(self.ctx, selection_set.ty);
+        let mut response_fields = BTreeMap::<ResponseEdge, ResponseValue>::new();
+        let fields = &selection_set.fields;
+
+        while let Some(key) = map.next_key::<Key<'_>>()? {
+            let key = key.as_ref();
+            let start = fields.partition_point(|field| &keys[field.expected_key] < key);
+
+            if start < fields.len() && &keys[fields[start].expected_key] == key {
+                let mut end = start + 1;
+                // All fields with the same expected_key (when aliases aren't support by upsteam)
+                while fields
+                    .get(end + 1)
+                    .map(|field| &keys[field.expected_key] == key)
+                    .unwrap_or_default()
+                {
+                    end += 1;
+                }
```

Comment:
Well could have been, slightly simplified in another PR to just compare the `expected_key` (u32) rather than `&str` as I'm comparing interned strings at that stage anyway. For the interned id I couldn't use `partition_point` again since it's sorted by the `&str` not their id